### ISSUE TITLE
Access component parts of probe name

### ISSUE THREAD

**Issue by ajor** [bpftrace/bpftrace#3145]
### Is your feature request related to a problem? Please describe.

When aggregating data from tracepoint probes, it is only currently possible to distinguish the data from each tracepoint by using the `probe` builtin:
```
tracepoint:syscalls:sys_enter_*
{
  @syscalls[probe] = count();
}
```
```
@syscalls[tracepoint:syscalls:sys_enter_close]: 35883
@syscalls[tracepoint:syscalls:sys_enter_openat]: 37641
@syscalls[tracepoint:syscalls:sys_enter_read]: 49833
@syscalls[tracepoint:syscalls:sys_enter_newfstatat]: 78194
```

It would be nice to have a way to access only part of the attach point's definition, so that we could generate output like this:
```
@syscalls[sys_enter_close]: 35883
@syscalls[sys_enter_openat]: 37641
@syscalls[sys_enter_read]: 49833
@syscalls[sys_enter_newfstatat]: 78194
```

### Describe the solution you'd like

Provide a way to access component parts of a probe's definition. `probe` will still return the full probe name, but individual pieces could be accessed as if it was an array:
```
probe;    // tracepoint:syscalls:sys_enter_read
probe[0]; // tracepoint
probe[1]; // syscalls
probe[2]; // sys_enter_read
```
Or tuple style:
```
probe.0; // tracepoint
probe.1; // syscalls
probe.2; // sys_enter_read
```

### Describe alternative solutions or features you've considered
`func` is an alternative for some use-cases (e.g. kprobes), but it doesn't work for all probe types (e.g. tracepoints) and its semantics are still not fully defined when tracing inlined functions.

### Problems
- I suppose this could only work if all attachment points of a probe were of the same type, i.e. all tracepoints or all USDTs. A mixture would lead to strange behaviour.
- What do we do about probes types which take a variable number of parameters? e.g. `kfunc:foo` vs `kfunc:module:foo`

**Comment by viktormalik** [comment]
> It would be nice to have a way to access only part of the attach point's definition, so that we could generate output like this:
> 
> ```
> @syscalls[sys_enter_close]: 35883
> @syscalls[sys_enter_openat]: 37641
> @syscalls[sys_enter_read]: 49833
> @syscalls[sys_enter_newfstatat]: 78194
> ```

Yeah, this is nice and useful. I can see use-cases for other probe types, too, e.g. grouping by kernel module for kfuncs.

> * What do we do about probes types which take a variable number of parameters? e.g. `kfunc:foo` vs `kfunc:module:foo`

This one will be tricky. Almost all probes have an optional part these days (kfuncs and kprobes have module, USDTs have namespace, uprobes has language, etc.) so using numerical indices may be quite confusing for users.

An alternative would be to introduce names for the individual parts but those would probably have to be different across various probe types. For example:
- tracepoints: `probe.type`, `probe.category`, `probe.event`
- kprobes/kfuncs: `probe.type`, `probe.module`, `probe.function`
- uprobes: `probe.type`, `probe.binary`, `probe.function`, `probe.offset`
- and so on...

The advantage is that we could go as far as interpret the probe part on the same "index" differently based on the field name - i.e. for uprobe, `probe.function` would return a string while `probe.offset` would return a number even though they are both the last part of the probe spec.

The disadvantage is that we'd have different part names for different probes (but since we wouldn't allow these for probes with attachpoints of mixed types, it wouldn't be that much of a problem). We could respect the naming from the manual so that it'd be clear to users what name to use.

**Comment by Phoenix500526** [comment]
I'm interesting in this issue. Plz assign it to me.

**Comment by jordalgo** [comment]
@Phoenix500526 I haven't seen a lot of interest in this issue TBH (@viktormalik and @ajor - feel free to advocate for this one). There might be higher priority things to work on if you're interested. How about any of these?
- https://github.com/bpftrace/bpftrace/issues/3924
- https://github.com/bpftrace/bpftrace/issues/4182

There is also this larger Project that probably needs more discussion:
- https://github.com/orgs/bpftrace/projects/6

I also plan to have a few more concrete items to work in the next two weeks if none of those are appealing.

**Comment by Arech8** [comment]
I'm also requesting the same ability in https://github.com/bpftrace/bpftrace/issues/4816. The only difference is basically that I requested a functional form instead of an array form (since there's already a function `probe()`), but I'm ok with any implementation **_as long as it exists_**!

I'm debugging an extremely fragile hang in a multithreaded application, and for that I need to dump as little information as possible. The biggest problem is that dumping a full `probe` output is very expensive, but I can't evade this for now. Having access to individual `probe` components would give me much more wiggle in dumping other vital data.

With respect to numeric indexes being confusing - maybe. But consider that they are by far the simplest and the faster solution to implement and have a zero maintenance cost compared to alternatives, b/c they are generic! Until a better solution is found, agreed upon between all stakeholders, and finally implemented, I'd much appreciate something simple AND working, such as numeric indices, that would unblock me. Users could live with simplified, but working option and bear responsibility to understand how it works.

**Comment by jordalgo** [comment]
> The biggest problem is that dumping a full probe output is very expensive

JC, what is the expensive part (just the memory used on the stack) ?

> With respect to numeric indexes being confusing - maybe. But consider that they are by far the simplest and the faster solution

We discussed this a bit in the last office hours and think this is not the right implementation due to the fact that probes have optional parts AND we're considering adding more elements to probes/attachpoints which would move these indexes and ultimately lead to breaking changes. I think we'd rather just provide properties on a `probe` object that you could access, e.g. `probe.function` or `probe.offset`.

**Comment by Arech8** [comment]
> what is the expensive part (just the memory used on the stack) ?

Additional CPU cycles. As I said, I'm debugging a very nasty hang in a complex multithreaded heterogeneous application. Just a tiny shift in operations latencies and the hang disappears. So far `bpftrace` was _the only_ tool I was able to find that can do some meaningful dumping that not always affects the hang. When the amount of data to dump increases, the hang disappears. I presume, because it take some more cycles to process and spew the log, and that might interfere with hang conditions.

**Comment by jordalgo** [comment]
I'm a little skeptical that logging the entire probe vs parts of it would have an effect here. That being said, there was some offline discussion about supporting named tuples which would open the door for having this functionality.  
- https://github.com/bpftrace/bpftrace/issues/4833

**Comment by jordalgo** [comment]
Since we now have Record types, this should be workable and similar to how `config` builtin was implemented: https://github.com/bpftrace/bpftrace/pull/5241

However, since we already have a `probe` builtin that returns the entire probe string, we'll probably need to go with something like `probe_info` for the new Record builtin. 

There will probably be some edge cases to handle/discuss around handling fields that are supported in one probe type but not another e.g. `offset`. Also probe glob expansion will be something we'll need to consider a bit.

**Comment by Rtoax** [comment]
Thank you very much. I am currently working [PR: Add probefunc builtin](https://github.com/bpftrace/bpftrace/pull/5300) on  a similar issue and look forward to seeing new discussions and progress on this topic.
